src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java

```diff
@@ -1494,6 +1488,125 @@ void loadIntoDataStore(
+
+        // see if the store exposes a structured grid coverage reader
+        // this is a ResourcePool reader, we should not close it
+        GridCoverageReader reader = store.getGridCoverageReader(null, null);
+        boolean isStructured = reader instanceof StructuredGridCoverage2DReader;
+
+        if (task.getUpdateMode() == UpdateMode.REPLACE && !isStructured) {
+            // Replacing the coverage
+            replaceCoverage(format, reader, store, task);
+        } else {
+            final String errorMessage =
+                    "Indirect raster import can only work against a structured grid coverage store (e.g., mosaic), this one is not: ";
+            if (!(isStructured)) {
+                throw new IllegalArgumentException(errorMessage + store);
+            }
+            StructuredGridCoverage2DReader sr = (StructuredGridCoverage2DReader) reader;
+            ImportData data = task.getData();
+            harvestImportData(sr, data);
+
+            // check we have a target resource, if not, create it
+            if (task.getUpdateMode() == UpdateMode.CREATE) {
+                if (task.getLayer() != null && task.getLayer().getId() == null) {
+                    addToCatalog(task);
+                }
+            }
+        }
+    }
+
+    private void replaceCoverage(
+            GridFormat format, GridCoverageReader reader, CoverageStoreInfo store, ImportTask task)
+            throws IOException {
+        // Get the imported data
+        ImportData data = task.getData();
+        if (!(data instanceof SpatialFile)) {
+            throw new IllegalArgumentException(
+                    "Only FileData are supported for replace. This is " + data);
+        }
+
+        SpatialFile fileData = (SpatialFile) data;
+        File file = fileData.getFile();
+        String newFile = file.toURI().toURL().getFile();
+
+        // When replacing, we need to make sure that the provided layer and store matches what we
+        // have in the catalog.
+        CoverageStoreInfo originalStore =
+                catalog.getCoverageStoreByName(store.getWorkspace().getName(), store.getName());
+        LayerInfo inputLayer = task.getLayer();
+        LayerInfo originalLayer = checkResourceExists(store, originalStore, inputLayer);
+        CoverageInfo originalCoverage = (CoverageInfo) originalLayer.getResource();
+
+        // Marking original names for restore
+        String name = originalCoverage.getName();
+        String nativeName = originalCoverage.getNativeName();
+
+        // Marking previous file for eventual delete
+        GridCoverage2DReader oldReader = ((GridCoverage2DReader) reader);
+        ServiceInfo readerInfo = oldReader.getInfo();
+        String oldFile = readerInfo.getSource().toURL().getFile();
+
+        // Setting up the new coverage
+        CatalogRepository repository = catalog.getResourcePool().getRepository();
+        Hints hints = new Hints(new RenderingHints(Hints.REPOSITORY, repository));
```

Comment:
These are nonstructured readers, why do we need to pass in the repository?

Reply:
good catch, I didn't properly clean up the code while checking the catalogBuilder usage. Fixing